banzai_nres/rv.py

```diff
@@ -0,0 +1,111 @@
+        correlation = kernel * flux
+        correlation /= variance
+
+        normalization = kernel * kernel / variance
+
+        # This gives us the signal to noise ratio in terms of N-sigma
+        snr = correlation.sum() / np.sqrt(normalization.sum())
+        x_cor.append(snr)
+
+    return np.array(x_cor)
+
+
+def barycentric_correction(time, exptime, ra, dec, site):
+    # TODO: figure out something better than hard-coding the NRES site coordinates
+    if site == 'lsc': lat, lon, height = '-30:10:2.64', '-70:48:17.28', 2198.
+    if site == 'elp': lat, lon, height = '30:40:12', '-104:01:12', 2070.
+    if site == 'cpt': lat, lon, height = '-32:22:48', '20:48:36', 1460.
+    if site == 'tlv': lat, lon, height = '30:35:45', '34:45:48', 875.
+    # The following is largely taken from the astropy example page
+    site_location = EarthLocation.from_geodetic(lat=lat, lon=lon, height=height*u.m)
+    sky_coordinates = SkyCoord(ra=ra*u.hourangle, dec=dec*u.deg)
+    # The time given in the NRES header is the exposure start time; correct this to the midpoint
+    # Eventually, we will want to use the flux-weighted mid-point calculated from emeter data
+    obs_time = Time(time, location=site_location) + exptime/2.*u.s
+    barycorr_rv = sky_coordinates.radial_velocity_correction(obstime=obs_time)
+    bjd_tdb = obs_time.tdb + obs_time.light_travel_time(sky_coordinates)
+    return barycorr_rv.to(u.m/u.s).value, bjd_tdb.to_value('jd')
+
+
+class RVCalculator(Stage):
+    TEMPLATE_FILENAME = pkg_resources.resource_filename('banzai_nres', 'data/g2v_template.fits')
+
+    def do_stage(self, image) -> ObservationFrame:
+        # Load in the template
+        template_hdu = fits.open(self.TEMPLATE_FILENAME)
+        ccfs = []
+        for i in image.fibers[image.fibers['fiber'] == image.science_fiber]['trace']:
+            # for steps in 1 km/s from -2000 to +2000 km/s
+            velocities = np.arange(-2000, 2001, 1)
+            # calculate the variance ahead of time
+            x_cor = cross_correlate(velocities, image.spectrum[i]['wavelength'], image.spectrum[i]['flux'],
+                                    image.spectrum[i]['uncertainty'],
+                                    template_hdu[1].data['wavelength'], template_hdu[1].data['flux'])
+
+            # take the peak
+            best_v = velocities[np.argmax(x_cor)]
+            # for steps of 1m/s around that peak take +- 2 km/s and repeat the process
+            velocities = np.arange(best_v - 2, best_v + 2 + 1e-4, 1e-3)
+            x_cor = cross_correlate(velocities, image.spectrum[i]['wavelength'], image.spectrum[i]['flux'],
+                                    image.spectrum[i]['uncertainty'],
+                                    template_hdu[1].data['wavelength'], template_hdu[1].data['flux'])
+
+            # Save the 1 m/s cross correlation function
+            ccfs.append({'v': velocities, 'xcor': x_cor})
```

Comment:
Consider breaking up calculation into : 
1. compute rough CCF on same grid for every order.
2. Sum the ccf's per order together (or average with weights or something). Take peak max of summed ccf as best_v
3. Recompute CCF around best_v +- delta v (same grid) for every order.
4. Save those CCF's per order. ccfs.append({'v': velocities, 'xcor': x_cor})